File: Programming languages/proyectos/rubycalc/lib/parser.rb
require 'ast'
require 'scanner'
require 'token'
require 'calcex'

class Parser
  def initialize(istream)
    @scan = Scanner.new istream
  end

  def parse
    prog
  end

  def prog
    result = expr
    t = @scan.getToken
    unless t.type == :eof
      #print "Expected EOF. Found ", t.type, ".\n"
      raise ParseError.new
    end
    result
  end

  def expr
    restExpr term
  end

  def restExpr(e)
    t = @scan.getToken
    case t.type
    when :add
      return restExpr(AddNode.new(e,term()))
    when :sub 
      return restExpr(SubNode.new(e,term()))
    end
    @scan.putBackToken
    e
  end

  def term
    restTerm storable
  end

  def restTerm(e)
    t = @scan.getToken
    case t.type
    when :times
      return restTerm(TimesNode.new(e,storable))
    when :divide
      return restTerm(DivideNode.new(e,storable))
    when :module
      return restTerm(ModNode.new(e,storable))
    end
    @scan.putBackToken
    e
  end

  def storable
    f = factor
    t = @scan.getToken
    if t.type == :keyword
      if t.lex == "S"
        return StoreNode.new f
      elsif t.lex == "P"
        return PlusNode.new f
      elsif t.lex == "M"
        return MinusNode.new f
      else
        #print "* parser error\n"
        raise ParseError.new
      end
    end
    @scan.putBackToken
    f
  end

  def factor
    t = @scan.getToken
    case t.type
    when :identifier
      key = t.lex.to_s
      return IdentifierNode.new(key)
    when :number 
      val = t.lex.to_i
      return NumNode.new(val)
    when :keyword
      if t.lex == "R"
         return RecallNode.new
      elsif t.lex == "C"
         return CleanNode.new
      elsif t.lex == "let"
         t = @scan.getToken
         unless t.type == :identifier
            #print "* parser error\n"
            raise ParseError.new
         end
         key = t.lex.to_s
         t = @scan.getToken
         unless t.type == :equals
            #print "* parser error\n"
            raise ParseError.new
         end     
         e = expr
         return AssignNode.new(e,key)
      else
        #print "* parser error\n"
        raise ParseError.new
      end
    when :lparen    
      e = expr
      t = @scan.getToken
      unless t.type == :rparent
        #print "* parser error\n"
        raise ParseError.new
      end
      return e
    else
      #print "* parser error\n"
      raise ParseError.new
    end
  end
  
end

File: Programming languages/proyectos/cppcalc/parser.cpp
#include "parser.h"
#include "calcex.h"
#include <string>
#include <sstream>

Parser::Parser(istream* in) { scan = new Scanner(in);}

Parser::~Parser() {
  try {
    delete scan;
  } catch (...) {
  }
}

AST* Parser::parse() { return prog(); }

AST* Parser::prog() {
  AST* result = expr();
  Token* t = scan->getToken();

  if (t->getType() != eof) {
    throw ParseError;
  }

  return result;
}

AST* Parser::expr() { return restExpr(term()); }

AST* Parser::restExpr(AST* e) {
  Token* t = scan->getToken();

  if (t->getType() == add) {
    return restExpr(new AddNode(e, term()));
  }

  if (t->getType() == sub) return restExpr(new SubNode(e, term()));

  scan->putBackToken();

  return e;
}

AST* Parser::term() { return restTerm(storable()); }

AST* Parser::restTerm(AST* e) {
  Token* t = scan->getToken();

  if (t->getType() == times) {
    return restTerm(new TimesNode(e, storable()));
  }

  if (t->getType() == divide) {
    return restTerm(new DivideNode(e, storable()));
  }

  if (t->getType() == mod) {
    return restTerm(new ModNode(e, storable()));
  }

  scan->putBackToken();

  return e;
}

AST* Parser::storable() { return memOperation(factor()); }

AST* Parser::memOperation(AST* e) {
  Token* t = scan->getToken();

  if (t->getType() == keyword) {
    if (t->getLex() == "S") {
      return new StoreNode(e);
    } else if (t->getLex() == "P") {
      return new PlusNode(e);
    } else if (t->getLex() == "M") {
      return new MinusNode(e);
    } else {
      throw ParseError;
    }
  }

  scan->putBackToken();

  return e;
}

AST* Parser::factor() {
  AST* result;
  Token* t = scan->getToken();

  if (t->getType() == number) {
    istringstream in(t->getLex());
    int val;
    in >> val;
    return new NumNode(val);
  }

  if (t->getType() == keyword) {
    if (t->getLex() == "R") {
      return new RecallNode();
    } else if (t->getLex() == "C") {
      return new CleanNode();
    } else if (t->getLex() == "let") {
      result = assignable();
      return result;
    } else {
      throw ParseError;
    }
  }

  if (t->getType() == lparen) {
    result = expr();
    t = scan->getToken();

    if (t->getType() == rparen)
      return result;
    else {
      //cout << "Expected )" << endl;
     // cout << "* parser error" << endl;
      throw ParseError;
    }
  }

  if (t->getType() == identifier) {
    string key = t->getLex();
    return new IdentifierNode(key);
  }

  //cout << "Expected number, R, C, (" << endl;
  //cout << "* parser error" << endl;
  throw ParseError;
}

AST* Parser::assignable() {
  Token* t = scan->getToken();
  if(t->getType() != identifier){
    throw ParseError;
  }
  string key = t->getLex();
  t = scan->getToken();
  if (t->getType() == equals) {
    return new AssignNode(expr(), key);
  } else {
    //cout << "Expected =" << endl;
   // cout << "* parser error" << endl;
    throw ParseError;
  }
}

